src/rules/selector-complexity.js

```diff
@@ -0,0 +1,526 @@
+		if (selector.type === "Combinator") {
+			if (compound.length > 0) {
+				compounds.push(compound);
+				compound = [];
+			}
+		} else {
+			compound.push(selector);
+		}
+	});
+
+	if (compound.length > 0) {
+		compounds.push(compound);
+	}
+
+	return compounds;
+}
+
+/**
+ * Get the location of a selector of a given name.
+ * @param {Array<Object>} allSelector All CSS selector nodes.
+ * @param {string} disallowedSelector The name of the disallowed selector.
+ * @returns {Object | undefined} The location of the disallowed selector, or undefined if not found.
+ */
+function getDisallowedSelectorsLocation(allSelector, disallowedSelector) {
+	return allSelector.find(selector => selector.name === disallowedSelector)
+		.loc;
+}
+
+/**
+ * An error for exceeding the maximum allowed selectors of a specific type.
+ * @param {Object} context The ESLint rule context object.
+ * @param {Object} selectorLoc The location of the selector.
+ * @param {number} maxValue The max number of selectors that are allowed.
+ * @returns {void}
+ */
+function exceedLimitError(context, selectorLoc, maxValue, selectorType) {
+	context.report({
+		loc: selectorLoc,
+		messageId: "maxSelectors",
+		data: {
+			selector: selectorType,
+			limit: String(maxValue),
+		},
+	});
+}
+
+/**
+ * Gives an array of CSS selectors of a specific type.
+ * @param {Array<Object>} selectors All CSS selectors nodes.
+ * @param {string} selectorType The type of CSS selector to filter out.
+ * @returns {Array<Object>} Filtered selectors.
+ */
+function getSelectors(selectors, selectorType) {
+	return selectors.filter(selector => selector.type === selectorType);
+}
+
+/**
+ * Get the names of all CSS selectors.
+ * @param {Array<Object>} selectors All CSS selector nodes.
+ * @returns {Array<string>} Array of selector names.
+ */
+function getSelectorNames(selectors) {
+	return selectors.map(selector => selector.name);
+}
+
+/**
+ * Get the location of the attribute matcher or operator in a given attribute selector.
+ * @param {Array<Object>} selectors All CSS selector nodes.
+ * @param {number} index The index of the attribute selector in the selectors array.
+ * @returns {{ startLoc: Object | undefined, endLoc: Object | undefined }} The start and end locations of the operator.
+ */
+function getOperatorLocation(selectors, index) {
+	const selector = selectors[index];
+	let startLoc;
+	let endLoc;
+
+	if (selector.name.type === "Identifier") {
+		startLoc = selector.name.loc.end;
+	}
+
+	if (selector.value) {
+		endLoc = selector.value.loc.start;
+	}
+
+	return { startLoc, endLoc };
+}
+
+/**
+ * Get the location of a given disallowed combinator.
+ * @param {Array<Object>} selectors All CSS selector nodes.
+ * @param {Array<Object>} combinatorNodes All combinator nodes.
+ * @param {string} combinator Name of combinator.
+ * @param {number} index The index of the given combinator.
+ * @returns {Object | undefined} The location of the disallowed combinator, or undefined if not found.
```

Comment:
The return value should not be `undefined` I think, otherwise an exception will be thrown.